### PR TITLE
perf(receipts): Receipt compression.

### PR DESCRIPTION
**Pull request by rakita**
poc for compression of data. It compresses receipts for around `x2.2`. 

Tests are done on sepolias first 3M blocks:
Plain receipts without compressions are `1.5gb` while compressed receipts are: `663.8mb`

```
Compressed:
|-------------------|-----------|--------------|------------|----------------|------------|
| CanonicalHeaders  | 3345845   | 45           | 20047      | 0              | 313.9 MiB  |
| HeaderTD          | 3345845   | 7            | 5318       | 0              | 83.2 MiB   |
| HeaderNumbers     | 3345845   | 66           | 15276      | 0              | 239.7 MiB  |
| Headers           | 3345845   | 446          | 215192     | 0              | 3.3 GiB    |
| BlockBodyIndices  | 3345845   | 7            | 4652       | 0              | 72.8 MiB   |
| BlockOmmers       | 49770     | 4            | 1862       | 0              | 29.2 MiB   |
| BlockWithdrawals  | 354923    | 14           | 11436      | 0              | 178.9 MiB  |
| TransactionBlock  | 1517533   | 4            | 2413       | 0              | 37.8 MiB   |
| Transactions      | 12528489  | 413          | 373760     | 655795         | 15.7 GiB   |
| TxHashNumber      | 48297     | 1            | 282        | 0              | 4.4 MiB    |
| Receipts          | 5953985   | 44           | 38354      | 4086           | 663.8 MiB  |
| PlainAccountState | 1036957   | 13           | 4156       | 0              | 65.1 MiB   |
| PlainStorageState | 7636174   | 2848         | 37046      | 0              | 623.3 MiB  |
| Bytecodes         | 58823     | 38           | 10974      | 31955          | 671.4 MiB  |
| AccountHistory    | 16913     | 1            | 424        | 0              | 6.6 MiB    |
| StorageHistory    | 1194      | 1            | 22         | 0              | 368 KiB    |
| AccountChangeSet  | 8450678   | 51           | 26423      | 0              | 413.7 MiB  |
| StorageChangeSet  | 28465943  | 8898         | 106753     | 0              | 1.8 GiB    |
| HashedAccount     | 1036957   | 22           | 5429       | 0              | 85.2 MiB   |
| HashedStorage     | 7636174   | 2869         | 36705      | 0              | 618.3 MiB  |
| AccountsTrie      | 74015     | 1            | 703        | 0              | 11 MiB     |
| StoragesTrie      | 518092    | 440          | 7622       | 0              | 126 MiB    |
| TxSenders         | 12528489  | 34           | 29137      | 0              | 455.8 MiB  |
| SyncStage         | 13        | 0            | 1          | 0              | 16 KiB     |
| SyncStageProgress | 3         | 0            | 1          | 0              | 16 KiB     |

Plain:
| Table Name        | # Entries | Branch Pages | Leaf Pages | Overflow Pages | Total Size |
|-------------------|-----------|--------------|------------|----------------|------------|
| CanonicalHeaders  | 3345845   | 45           | 20047      | 0              | 313.9 MiB  |
| HeaderTD          | 3345845   | 7            | 5318       | 0              | 83.2 MiB   |
| HeaderNumbers     | 3345845   | 66           | 15276      | 0              | 239.7 MiB  |
| Headers           | 3345845   | 446          | 215192     | 0              | 3.3 GiB    |
| BlockBodyIndices  | 3345845   | 7            | 4652       | 0              | 72.8 MiB   |
| BlockOmmers       | 49770     | 4            | 1862       | 0              | 29.2 MiB   |
| BlockWithdrawals  | 354923    | 14           | 11436      | 0              | 178.9 MiB  |
| TransactionBlock  | 1517533   | 4            | 2413       | 0              | 37.8 MiB   |
| Transactions      | 12528489  | 413          | 373760     | 655795         | 15.7 GiB   |
| TxHashNumber      | 5953985   | 89           | 25537      | 0              | 400.4 MiB  |
| Receipts          | 5953985   | 77           | 68916      | 32347          | 1.5 GiB    |
| PlainAccountState | 1036957   | 13           | 4156       | 0              | 65.1 MiB   |
| PlainStorageState | 7636174   | 2848         | 37046      | 0              | 623.3 MiB  |
| Bytecodes         | 58823     | 38           | 10974      | 31955          | 671.4 MiB  |
| AccountHistory    | 16913     | 1            | 424        | 0              | 6.6 MiB    |
| StorageHistory    | 561368    | 62           | 10397      | 0              | 163.4 MiB  |
| AccountChangeSet  | 8450678   | 51           | 26423      | 0              | 413.7 MiB  |
| StorageChangeSet  | 28465943  | 8898         | 106753     | 0              | 1.8 GiB    |
| HashedAccount     | 1036957   | 22           | 5429       | 0              | 85.2 MiB   |
| HashedStorage     | 7636174   | 2869         | 36705      | 0              | 618.3 MiB  |
| AccountsTrie      | 86347     | 1            | 823        | 0              | 12.9 MiB   |
| StoragesTrie      | 606838    | 514          | 8846       | 0              | 146.2 MiB  |
| TxSenders         | 12528489  | 34           | 29137      | 0              | 455.8 MiB  |
| SyncStage         | 13        | 0            | 1          | 0              | 16 KiB     |
| SyncStageProgress | 3         | 0            | 1          | 0              | 16 KiB     |
```

Additional time of compression is insignificant:
```
Plain:
2023-04-23T16:06:58.169299Z  INFO Stage committed progress stage=Execution checkpoint=1000000
2023-04-23T16:13:39.837735Z  INFO Stage finished executing stage=MerkleExecute checkpoint=3000000
401s

Compressed:
2023-04-23T15:56:30.099959Z  INFO Executing stage stage=Execution from=1000000
2023-04-23T16:03:18.529737Z  INFO Stage committed progress stage=Execution checkpoint=3000000
408s
```

### CODE REVIEW

**Comment by onbjerg** [comment]
Does it make sense to compress the receipts at all given we eventually have to move blooms and logs into their own tables? That would reduce the overhead we are seeing from overflow pages

File: crates/storage/db/src/tables/compressed.rs [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+    type Key = T::Key;
+
+    type Value = CompressedValue<T::Value>;
+}
+
+
+
+/// Raw table value.
+#[derive(Debug, Clone, PartialEq, Eq, PartialOrd, Serialize, Ord, Hash)]
+pub struct CompressedValue<V: Value> {
```
Comment (gakonst) [inline crates/storage/db/src/tables/compressed.rs:26]: I would call this ZstdCompressed<T> perhaps

File: crates/storage/provider/src/providers/mod.rs [hunk 2]
```diff
@@ -289,38 +289,40 @@ impl<DB: Database> TransactionsProvider for ShareableDatabase<DB> {
 
 impl<DB: Database> ReceiptProvider for ShareableDatabase<DB> {
     fn receipt(&self, id: TxNumber) -> Result<Option<Receipt>> {
-        self.db.view(|tx| tx.get::<tables::Receipts>(id))?.map_err(Into::into)
+        Ok(None)
```
Comment (gakonst) [inline crates/storage/provider/src/providers/mod.rs:292]: Revert?

**Review by gakonst — COMMENTED** [review]
Nice. We should also try this for the transactions table

**Comment by rakita** [comment]
@onbjerg @gakonst This was just a simple PoC to show the effect of generic compression, we still need to check what kind of compression is best for our use case and what tables we want to compress.

Maybe those tables are just tx and receipts, maybe all other tables except for some ( Merkle tree as they are hashes, and plain state as we need very fast access?) are good to compress. 

About naming, it will depend on the final implementation (or even compression lib that we use), maybe it is possible (good idea) to embed this inside `main_codec` macro that we have. pinging @joshieDo

**Comment by rakita** [comment]
Closing this as other work on this topic is in progress.

**Comment by gakonst** [comment]
yep - ref #2483

**Pull request closed without merging**